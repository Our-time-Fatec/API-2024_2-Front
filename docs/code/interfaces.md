
---
title: interfaces
description: 'Pasta que contém as definições de interfaces utilizadas no projeto.'
---
# interfaces

A pasta `interfaces` contém as definições de interfaces que são utilizadas em todo o projeto. As interfaces são fundamentais para garantir a tipagem e a estrutura dos dados, proporcionando uma melhor organização e manutenção do código.

## Estrutura da Pasta

Abaixo estão as interfaces disponíveis nesta pasta:

- **IAlimento.ts**: Define a estrutura de um alimento, incluindo suas propriedades e tipos.
- **ICategoria.ts**: Define a estrutura de uma categoria de alimentos.
- **IDieta.ts**: Define a estrutura de uma dieta, incluindo os alimentos e categorias associadas.
- **IGrafico.ts**: Define a estrutura de dados para gráficos utilizados no aplicativo.
- **ILogin.ts**: Define a estrutura de dados para o processo de login.
- **IUsuario.ts**: Define a estrutura de um usuário, incluindo suas informações e preferências.

Essas interfaces são utilizadas em diversos componentes e serviços do projeto, garantindo a consistência e a segurança dos dados manipulados.
