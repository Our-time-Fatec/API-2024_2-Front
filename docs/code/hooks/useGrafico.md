---
title: useGrafico
description: 'Um hook personalizado para gerenciar e obter dados da dieta semanal, incluindo cache e requisições à API.'
---

# useGrafico

O `useGrafico` é um hook personalizado que permite gerenciar e obter dados sobre a dieta semanal de um usuário. Ele utiliza o armazenamento assíncrono para cache e realiza requisições à API para buscar informações atualizadas.

## Estrutura do Hook

### Estado

O hook mantém três estados principais:

- `dietaSemanal`: Armazena os dados da dieta semanal, que podem ser do tipo `IDietaSemanal` ou `null`.
- `loading`: Um booleano que indica se os dados estão sendo carregados.
- `error`: Uma string que armazena mensagens de erro, caso ocorram durante a requisição.

### Funções

#### fetchAlimentosConsumidosSemana

Esta função é responsável por buscar os dados da dieta semanal. Ela realiza as seguintes etapas:

1. **Cache**: Tenta obter os dados do cache usando `AsyncStorage`. Se os dados estiverem disponíveis e `refreshTrigger` não for verdadeiro, os dados em cache são utilizados.
2. **Requisição à API**: Se os dados não estiverem em cache ou se `refreshTrigger` for verdadeiro, a função faz uma requisição à API para obter os dados mais recentes.
3. **Atualização do Estado**: Após obter os dados, atualiza o estado `dietaSemanal` e armazena os dados no cache.
4. **Tratamento de Erros**: Em caso de erro, uma mensagem de erro é definida no estado.

### Efeitos

O hook utiliza o `useEffect` para chamar `fetchAlimentosConsumidosSemana` sempre que o hook é montado ou quando `refreshTrigger` muda.

## Retorno

O hook retorna um objeto com as seguintes propriedades:

- `dietaSemanal`: Os dados da dieta semanal.
- `loading`: O estado de carregamento.
- `error`: Mensagem de erro, se houver.
- `refreshGrafico`: Uma função que pode ser chamada para atualizar os dados manualmente.

## Exemplo de Uso

```javascript
import useGrafico from './hooks/useGrafico';

const MyComponent = () => {
  const { dietaSemanal, loading, error, refreshGrafico } = useGrafico();

  // Lógica do componente...
};
```

## Considerações Finais

O `useGrafico` é uma solução eficiente para gerenciar dados da dieta semanal, permitindo que os desenvolvedores integrem facilmente a funcionalidade de cache e requisições à API em seus componentes React.