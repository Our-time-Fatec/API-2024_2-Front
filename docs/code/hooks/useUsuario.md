---
title: useUsuario
description: 'Um hook personalizado para gerenciar o estado do usuário, incluindo carregamento e tratamento de erros.'
---

# useUsuario

O `useUsuario` é um hook personalizado que permite gerenciar o estado do usuário em uma aplicação React Native. Ele fornece funcionalidades para buscar informações do usuário, controlar o estado de carregamento e lidar com erros.

## Estrutura do Hook

O hook utiliza os seguintes estados:

- `usuario`: Armazena as informações do usuário, que são do tipo `IUsuario` ou `null`.
- `loading`: Um booleano que indica se os dados estão sendo carregados.
- `error`: Uma string que armazena mensagens de erro, caso ocorram.

## Funcionalidades

### fetchUsuario

A função `fetchUsuario` é responsável por buscar as informações do usuário. Ela realiza uma requisição GET para o endpoint `/usuario/mydetails` utilizando a função `requestWithRefresh`. 

- **Tratamento de Erros**: Caso ocorra um erro durante a requisição, a mensagem de erro é armazenada no estado `error`.

### useEffect

O hook utiliza o `useEffect` para chamar `fetchUsuario` assim que o componente que utiliza o hook é montado. Isso garante que as informações do usuário sejam carregadas automaticamente.

## Retorno

O `useUsuario` retorna um objeto contendo:

- `usuario`: As informações do usuário.
- `loading`: O estado de carregamento.
- `error`: Mensagem de erro, se houver.
- `refreshUser`: Uma função para atualizar as informações do usuário chamando `fetchUsuario` novamente.

## Exemplo de Uso

```javascript
import useUsuario from 'hooks/useUsuario';

const MeuComponente = () => {
    const { usuario, loading, error, refreshUser } = useUsuario();

    if (loading) return <Text>Carregando...</Text>;
    if (error) return <Text>{error}</Text>;

    return (
        <View>
            <Text>Nome: {usuario?.nome}</Text>
            <Button title="Atualizar" onPress={refreshUser} />
        </View>
    );
};
```

## Considerações Finais

O `useUsuario` é uma solução eficiente para gerenciar o estado do usuário em aplicações React Native, facilitando a integração com APIs e o tratamento de estados de carregamento e erro.