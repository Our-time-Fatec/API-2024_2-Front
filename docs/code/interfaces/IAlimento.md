---
title: IAlimento
description: 'Interface que define a estrutura de um alimento, incluindo suas propriedades nutricionais e informações adicionais.'
---

# IAlimento

A interface `IAlimento` descreve a estrutura de um alimento, incluindo suas propriedades nutricionais e informações adicionais. Esta interface é utilizada para garantir que os objetos de alimento sigam um formato consistente em toda a aplicação.

## Estrutura da Interface

### IAlimento

```typescript
interface IAlimento {
    _id?: string;                // Identificador único do alimento (opcional)
    nome: string;                // Nome do alimento
    preparo: string;             // Modo de preparo do alimento
    categoriaUrl?: string;       // URL da categoria do alimento (opcional)
    porcao: number;              // Tamanho da porção em gramas
    quantidade?: number;         // Quantidade disponível (opcional)
    categoriaNome?: string;      // Nome da categoria do alimento (opcional)
    categoriaCodigo: number;     // Código da categoria do alimento
    detalhes: Detalhes;          // Detalhes nutricionais do alimento
    criadoEm?: Date;             // Data de criação do alimento (opcional)
}
```

### Detalhes

```typescript
interface Detalhes {
    valorEnergetico: number;     // Valor energético em kcal
    proteinas: number;           // Quantidade de proteínas em gramas
    carboidratos: number;        // Quantidade de carboidratos em gramas
    fibras: number;              // Quantidade de fibras em gramas
    lipidios: number;            // Quantidade de lipídios em gramas
}
```

## Exportação

A interface `IAlimento` e a interface `Detalhes` são exportadas para serem utilizadas em outras partes da aplicação.

```typescript
export { IAlimento, Detalhes };
```