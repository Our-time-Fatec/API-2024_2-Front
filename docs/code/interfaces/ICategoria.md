---
title: ICategoria
description: 'Interface que define a estrutura de uma categoria no sistema.'
---

# ICategoria

A interface `ICategoria` define a estrutura de uma categoria no sistema, incluindo suas propriedades essenciais.

## Propriedades

- **_id**: `string`  
  Identificador único da categoria.

- **nome**: `string`  
  Nome da categoria.

- **codigo**: `number`  
  Código numérico associado à categoria.

- **urlPlaceholder**: `string`  
  URL de um placeholder relacionado à categoria.

## Exportação

A interface é exportada como padrão para ser utilizada em outras partes do sistema:

```typescript
export default ICategoria;
```