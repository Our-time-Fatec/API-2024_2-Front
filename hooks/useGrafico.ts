import { useState, useEffect, useCallback } from 'react';
import { requestWithRefresh } from '../services/api';
import { IDietaSemanal } from '../interfaces/IGrafico';

const useGrafico = () => {
    const [dietaSemanal, setDietaSemanal] = useState<IDietaSemanal | null>(null);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState<string | null>(null);

    const fetchAlimentosConsumidosSemana = useCallback(async () => {
        setLoading(true);
        setError(null);
        try {
            const response = await requestWithRefresh({
                method: 'GET',
                url: '/alimentoConsumido/me/semana', // URL do seu endpoint
            });

            const dietaSemanalData: IDietaSemanal = response.data;

            // Atualiza o estado com os dados completos, sem filtros
            console.log(JSON.stringify(dietaSemanalData))
            setDietaSemanal(dietaSemanalData);
        } catch (err) {
            console.error('Erro ao buscar alimentos consumidos na semana:', err);
            setError('Erro ao buscar alimentos consumidos na semana.');
        } finally {
            setLoading(false);
        }
    }, []);

    useEffect(() => {
        fetchAlimentosConsumidosSemana()
    }, [fetchAlimentosConsumidosSemana]);

    const refreshAlimentosConsumidos = useCallback(() => {
        fetchAlimentosConsumidosSemana();
    }, [fetchAlimentosConsumidosSemana]);

    return {
        dietaSemanal,
        loading,
        error,
        refreshAlimentosConsumidos,
    };
};

export default useGrafico;
