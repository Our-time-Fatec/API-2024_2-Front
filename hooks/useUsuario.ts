import { useState, useEffect, useCallback } from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { IUsuario } from "../interfaces/IUsuario";
import { requestWithRefresh } from "../services/api"; // Ajuste o caminho conforme necessário

const useUsuario = (forcedRefresh?: any) => {
  const [usuario, setUsuario] = useState<IUsuario | null>(null);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);

  const fetchUsuario = useCallback(
    async (forcedRefresh?: any) => {
      setLoading(true);
      setError(null);

      try {
        // Primeiro, tente obter os dados do usuário do cache
        if (!forcedRefresh) {
          const cachedUsuario = await AsyncStorage.getItem("usuario");
          if (cachedUsuario) {
            setUsuario(JSON.parse(cachedUsuario));
            setLoading(false);
            return;
          }
        }

        // Se não houver dados no cache, faça a requisição à API
        const response = await requestWithRefresh({
          method: "GET",
          url: "/usuario/mydetails",
        });
        setUsuario(response.data);

        // Armazene os dados no cache
        await AsyncStorage.setItem("usuario", JSON.stringify(response.data));
      } catch (err: any) {
        setError(err.message || "Erro ao buscar informações do usuário");
      } finally {
        setLoading(false);
      }
    },
    [forcedRefresh]
  );

  useEffect(() => {
    fetchUsuario();
  }, [fetchUsuario]);

  return { usuario, loading, error, refreshUser: fetchUsuario };
};

export default useUsuario;
