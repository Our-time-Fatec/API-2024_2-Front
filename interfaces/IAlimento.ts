interface Detalhes {
    valorEnergetico: number;
    proteinas: number;
    carboidratos: number;
    fibras: number;
    lipidios: number;
}

interface IAlimento {
    _id?: string;
    nome: string;
    preparo: string;
    categoriaUrl?: string;
    porcao: number;
    categoriaNome?: string;
    categoriaCodigo: number;
    detalhes: Detalhes;
}

export { IAlimento, Detalhes };