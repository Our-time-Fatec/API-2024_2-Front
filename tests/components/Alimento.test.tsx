import React from 'react';
import TestRenderer from 'react-test-renderer';
import AlimentoItem from '../../components/alimento'; // ajuste o caminho conforme necessário
import { IAlimento } from '../../interfaces/IAlimento';
import useGrafico from '../../hooks/useGrafico';
import useUsuario from '../../hooks/useUsuario';

jest.mock("../../hooks/useGrafico", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("../../hooks/useUsuario", () => ({
  __esModule: true,
  default: jest.fn(),
}));

const mockAlimento: IAlimento = {
    _id: "1",
    nome: "Banana",
    porcao: 100,
    preparo: "In Natura",
    categoriaNome: "Frutas",
    categoriaUrl: "",
    criadoEm: new Date(),
    quantidade: 2,
    nomeGrupo: "Café da Manhã",
    detalhes: {
        valorEnergetico: 89,
        carboidratos: 22.8,
        proteinas: 1.1,
        fibras: 2.6,
        lipidios: 0.3,
    },
    categoriaCodigo: 0
};

describe('Componente AlimentoItem', () => {
  beforeEach(() => {
    (useGrafico as jest.Mock).mockReturnValue({
      dietaSemanal: [],
      refreshGrafico: jest.fn(),
    });

    (useUsuario as jest.Mock).mockReturnValue({
      usuario: {},
      refreshUser: jest.fn(),
    });
  });

  it('Deve renderizar corretamente com os dados do alimento fornecidos', () => {
    const testRenderer = TestRenderer.create(
      <AlimentoItem
        alimento={mockAlimento}
        isUserAlimento={true}
        onEdit={jest.fn()}
        onDelete={jest.fn()}
      />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
