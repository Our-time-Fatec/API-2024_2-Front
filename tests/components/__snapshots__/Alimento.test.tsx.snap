// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente AlimentoItem Deve renderizar corretamente com os dados do alimento fornecidos 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "borderColor": "#ddd",
      "borderRadius": 8,
      "borderWidth": 1,
      "elevation": 2,
      "flex": 1,
      "justifyContent": "space-between",
      "margin": 8,
      "overflow": "hidden",
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 4,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f0f0f0",
        "height": 100,
        "justifyContent": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
    testID="imageTouchable"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#e0e0e0",
          "borderRadius": 8,
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        }
      }
      testID="imagePlaceholder"
    >
      <Text
        style={
          {
            "color": "#888",
            "fontSize": 14,
          }
        }
      >
        Imagem não disponível
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      style={
        {
          "padding": 8,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 16,
            "fontWeight": "bold",
            "marginBottom": 4,
          }
        }
        testID="alimentoNome"
      >
        Banana
      </Text>
      <Text
        style={
          {
            "color": "#555",
            "fontSize": 14,
            "marginBottom": 2,
          }
        }
        testID="alimentoPorcao"
      >
        100g
      </Text>
      <Text
        style={
          {
            "color": "#555",
            "fontSize": 14,
            "marginBottom": 2,
          }
        }
        testID="alimentoPreparo"
      >
        Preparo: 
        In Natura
      </Text>
      <Text
        style={
          {
            "color": "#555",
            "fontSize": 14,
            "marginBottom": 2,
          }
        }
        testID="alimentoCategoria"
      >
        Categoria: 
        Frutas
      </Text>
    </View>
    <View>
      <View
        style={
          {
            "flexDirection": "column",
            "gap": 3,
            "justifyContent": "space-between",
            "marginBottom": 10,
            "paddingHorizontal": 10,
            "paddingVertical": 1,
            "width": "100%",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#2d74da",
              "borderRadius": 5,
              "flex": 0.5,
              "marginTop": 5,
              "opacity": 1,
              "paddingVertical": 10,
            }
          }
          testID="editButton"
        >
          <Text
            style={
              {
                "color": "#fff",
                "fontSize": 16,
              }
            }
          >
            Editar
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#e13f2f",
              "borderRadius": 5,
              "flex": 0.5,
              "marginTop": 5,
              "opacity": 1,
              "paddingVertical": 10,
            }
          }
          testID="deleteButton"
        >
          <Text
            style={
              {
                "color": "#fff",
                "fontSize": 16,
              }
            }
          >
            Remover
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "column",
            "gap": 3,
            "justifyContent": "space-between",
            "marginBottom": 10,
            "paddingHorizontal": 10,
            "paddingVertical": 1,
            "width": "100%",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#2d74da",
              "borderRadius": 5,
              "flex": 0.5,
              "marginTop": 5,
              "opacity": 1,
              "paddingVertical": 10,
            }
          }
          testID="consumirButton"
        >
          <Text
            style={
              {
                "color": "#fff",
                "fontSize": 16,
              }
            }
          >
            Consumir
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
